File: sdk/bundled/src/main/java/ink/reactor/sdk/bundled/config/AbstractConfigService.java
package ink.reactor.sdk.bundled.config;

import ink.reactor.sdk.config.ConfigService;
import ink.reactor.sdk.config.section.ConfigSection;

import java.io.FileNotFoundException;
import java.io.IOException;
import java.io.InputStream;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.StandardCopyOption;

public abstract class AbstractConfigService implements ConfigService {

    @Override
    public ConfigSection createIfAbsentAndLoad(final String fileName, final Path outDestination, final ClassLoader classLoader) throws IOException {
        if (Files.exists(outDestination)) {
            return load(outDestination);
        }

        try(final InputStream inputStream = classLoader.getResourceAsStream(fileName)) {
            if (inputStream == null) {
                throw new FileNotFoundException("Can't found the file " + fileName + " inside of the jar " + classLoader);
            }
            Files.copy(inputStream, outDestination, StandardCopyOption.REPLACE_EXISTING);
            return load(outDestination);
        }
    }
}

File: sdk/bundled/src/main/kotlin/ink/reactor/sdk/bundled/config/AbstractConfigService.kt
package ink.reactor.sdk.bundled.config

import ink.reactor.sdk.config.ConfigSection
import ink.reactor.sdk.config.ConfigService
import java.io.FileNotFoundException
import java.io.IOException
import java.io.InputStream
import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.StandardCopyOption
import kotlin.io.path.createDirectories
import kotlin.io.path.exists

abstract class AbstractConfigService : ConfigService {

    @Throws(IOException::class)
    override fun createIfAbsentAndLoad(
        fileName: String,
        outDestination: Path,
        classLoader: ClassLoader
    ): ConfigSection {
        if (outDestination.exists()) {
            return load(outDestination)
        }

        val inputStream: InputStream = classLoader.getResourceAsStream(fileName)
            ?: throw FileNotFoundException("Can't find the file '$fileName' inside of the jar.")

        inputStream.use { input ->
            outDestination.parent?.createDirectories()
            Files.copy(input, outDestination, StandardCopyOption.REPLACE_EXISTING)
        }

        return load(outDestination)
    }
}
